Normalize image names in prefix substitutor

findepi opened the pull request:
Optionally normalize docker image names to use `library/` prefix when they don't have any namespace at all. This allows for the `PrefixingImageNameSubstitutor` to be used to redirect Docker Hub requests to a private registry without having to apply otherwise redundant changes to the test code.

<!--
Thanks for contributing to Testcontainers. Please review the following notes before
submitting a pull request.

New Modules:

If you are contributing a new module, please add your module name to the following files:

* `./.github/ISSUE_TEMPLATE/bug_report.yaml`
* `./.github/ISSUE_TEMPLATE/enhancement.yaml`
* `./.github/ISSUE_TEMPLATE/feature.yaml`
* `./.github/dependabot.yml`
* `./.github/labeler.yml`

Also make sure that your new module has the appropriate documentation under `./docs/modules/`

Before committing any change, please run `./gradlew checkstyleMain checkstyleTest spotlessApply` and fix any issues that occur.

Dependency Upgrades:
Please do not open a pull request to update only a version dependency. Existing process will perform
the upgrade every week. However, if the upgrade involves more changes (changes for deprecated API) then
your pull request is very welcome.

Describing Your Changes:

If, after having reviewed the notes above, you're ready to submit your pull request, please
provide a brief description of the proposed changes and their context. If they fix a bug, please
describe the broken behaviour and how the changes fix it. If they make an enhancement,
please describe the new functionality and why you believe it's useful. If your pull
request relates to any existing issues, please reference them by using the issue number
prefixed with #.
-->

eddumelendez commented:
@findepi thanks for your contribution, can you please add some docs?

findepi commented:
@eddumelendez thanks for review!
sure, where this should be documented?

eddumelendez commented:
I think https://java.testcontainers.org/features/image_name_substitution/ is ok

findepi commented on docs/features/image_name_substitution.md:
```diff
@@ -77,6 +77,13 @@ Testcontainers will not apply the prefix to:
 * non-Hub image names (e.g. where another registry is set)
 * Docker Hub image names where the hub registry is explicitly part of the name (i.e. anything with a `docker.io` or `registry.hub.docker.com` host part)
 
+If you want your registry to handle both official Docker Hub images (e.g `postgres`)
+as well as images from other registries (e.g `mycompany/postgres`), you can use the
+`TESTCONTAINERS_HUB_IMAGE_NAME_NORMALIZE` environment variable or the `hub.image.name.normalize`
+configuration option. When set to `true`, Testcontainers will normalize the official Docker Hub
+image names to start with `library/`. When this option is used, Testcontainers will additionally
+disable image compatibility checks done by some containers, so it the compatibility responsibility
+is on the user.
```
Added this. cc @eddumelendez.

Did i get the env variable name right?
There is automatic conversion from TESTCONTAINERS_HUB_IMAGE_NAME_NORMALIZE to hub.image.name.normalize, right?

eddumelendez replied:
yes, the name is correct.

> When this option is used, Testcontainers will additionally disable image compatibility checks done by some containers, so it the compatibility responsibility is on the user.

I don't think this is the case because the check compatibility is executed first. Can you please add a test similar to https://github.com/testcontainers/testcontainers-java/blob/00cc0ea84b73642e3494e6b7a0384a5e3fa201bd/core/src/test/java/org/testcontainers/utility/ImageNameSubstitutorTest.java#L58-L68 in order to cover it?

findepi replied:
> I don't think this is the case because the check compatibility is executed first.

that makes sense!

however, when testing this PR internally, i run into some issues until i added "as compatible substitute" call.
is it possible that there are some additional checks somewhere else? sadly, don't have stacktrace now

findepi commented:
The build failed ([core](https://github.com/testcontainers/testcontainers-java/actions/runs/8880952232/job/24382176159?pr=8509#logs))

```
Gradle Test Executor 10 > org.testcontainers.junit.DockerComposeContainerScalingTest > simpleTest FAILED
    org.junit.ComparisonFailure: [Each redis instance is separate] expected:<"[1]"> but was:<"[2]">
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at org.testcontainers.junit.DockerComposeContainerScalingTest.simpleTest(DockerComposeContainerScalingTest.java:54)
```

I will need some guidance, I don't know how to fix it.

eddumelendez commented:
@findepi can you run `./gradlew spotlessApply --no-daemon`, please?

findepi commented:
@eddumelendez done!

findepi commented:
CI green this time. Sorry for missing spotless on the previous push.

eddumelendez commented on core/src/main/java/org/testcontainers/utility/DockerImageName.java:
```diff
@@ -26,7 +26,7 @@ public final class DockerImageName {
     private static final Pattern REPO_NAME = Pattern.compile(REPO_NAME_PART + "(/" + REPO_NAME_PART + ")*");
 
-    private static final String LIBRARY_PREFIX = "library/";
+    public static final String LIBRARY_PREFIX = "library/";
```
```suggestion
    static final String LIBRARY_PREFIX = "library/";
```

findepi replied:
done!

findepi commented:
I see test failure https://app.circleci.com/pipelines/github/testcontainers/testcontainers-java/11624/workflows/67da8da7-fbdc-42d2-94ab-c6e59c101324/jobs/43677

```
     Caused by:
            com.github.dockerjava.api.exception.DockerClientException: Could not pull image: [DEPRECATION NOTICE] Docker Image Format v1 and Docker Image manifest version 2, schema 1 support is disabled by default and will be removed in an upcoming release. Suggest the author of docker.io/library/redis:3.0.2 to upgrade the image to the OCI Format or Docker Image manifest v2, schema 2. More information at https://docs.docker.com/go/deprecated-image-specs/
```

in https://github.com/trinodb/trino we solved similar failures by bumping the image version, but I am not sure this is the way to go here.

@eddumelendez please let me know what do you think about this

findepi committed:
Normalize image names in prefix substitutor

Optionally normalize docker image names to use `library/` prefix when
they don't have any namespace at all. This allows for the
`PrefixingImageNameSubstitutor` to be used to redirect Docker Hub
requests to a private registry without having to apply otherwise
redundant changes to the test code.

rp-arielrodriguez commented:
Any plans for this PR?
We’re currently using Docker ECR Pull Through Cache (PTC) in our CI, and had to create a custom ImageNameSubstitutor to take advantage of caching with Docker Hub images.
It would be great if something like this could be supported out of the box.
